src/hotspot/cpu/s390/c1_Runtime1_s390.cpp

```diff
@@ -589,6 +589,76 @@ OopMapSet* Runtime1::generate_code_for(C1StubId id, StubAssembler* sasm) {
       __ z_br(Z_R14);
     }
     break;
+    case C1StubId::is_instance_of_id:
+    {
+      // Mirror: Z_ARG1(R2)
+      // Object: Z_ARG2
+      // Temps: Z_ARG3, Z_ARG4, Z_ARG5, Z_R10, Z_R11
+      // Result: Z_RET(R2)
+
+      // Get the Klass* into Z_ARG3
+      Register klass = Z_ARG3 , obj = Z_ARG2, result = Z_RET;
+      Register temp0 = Z_ARG4, temp1 = Z_ARG5, temp2 = Z_R10, temp3 = Z_R11;
+
+      __ z_lg(klass, Address(Z_ARG1, java_lang_Class::klass_offset()));
+
+      Label is_secondary;
+
+      __ clear_reg(result /* Z_R2 */, true /* whole_reg */, false /* set_cc */);  // sets result=0 (failure)
+
+      __ z_ltgr(klass, klass); // Klass is null
+      __ z_bcr(Assembler::bcondEqual, Z_R14);
+
+      __ z_ltgr(obj, obj); // obj is null
+      __ z_bcr(Assembler::bcondEqual, Z_R14);
+
+      __ z_llgf(temp0, Address(klass, in_bytes(Klass::super_check_offset_offset())));
+      __ compare32_and_branch(temp0, in_bytes(Klass::secondary_super_cache_offset()), Assembler::bcondEqual, is_secondary); // Klass is a secondary superclass
+
+      // Klass is a concrete class
+      __ load_klass(temp1, obj);
+      __ z_cg(klass, Address(temp1, temp0));
+
+      // result is already holding 0, denoting NotEqual case
+      __ load_on_condition_imm_32(result, 1, Assembler::bcondEqual);
+      __ z_br(Z_R14);
+
+      __ bind(is_secondary);
+
+      __ load_klass(obj, obj);
+
+      // This is necessary because I am never in my own secondary_super list.
+      __ z_cgr(obj, klass);
+      __ load_on_condition_imm_32(result, 1, Assembler::bcondEqual);
+      __ z_bcr(Assembler::bcondEqual, Z_R14);
+
+      const int frame_size = 2*BytesPerWord + frame::z_abi_160_size;
+      __ save_return_pc();
+      __ push_frame(frame_size);
+
+      // Z_R10 and Z_R11 are caller saved, so we must push them before any use
+      __ z_stg(temp2 /*Z_R10*/, 0*BytesPerWord + frame::z_abi_160_size, Z_SP);
+      __ z_stg(temp3 /*Z_R11*/, 1*BytesPerWord + frame::z_abi_160_size, Z_SP);
+      assert(2*BytesPerWord + frame::z_abi_160_size == frame_size, "check");
+
```

Comment:
I think you may be able temporarily to save R10 and R11 in the floating-point registers. You have plenty of call-clobbered FP registers, I think. This might work better than creating a stack frame. I guess it's possible to copy from an integer register to a floating-point register without altering any of the bits. There might be no performance advantage, but I think that it's worth a try.

Reply:
I see some further improvements. I will post the result down.